### PR TITLE
Update several tool-info modules to BaseTool2

### PR DESCRIPTION
**Pull request by skanav**
updated tool info modules of verification (and validation) tools from sv-comp 20 to start using data models from the task dict.
this is required for the competition after [PR 169](https://github.com/sosy-lab/sv-comp/pull/169) in benchmark definitions repo is merged

The following are still remaining: 
1. ultimate based tools - it is a bit more complicated. Should be dealt separately

### CODE REVIEW

**Comment by danieldietsch** [comment]
Should I update the Ultimate tools? How soon is it needed?

File: benchexec/tools/brick.py
Comment (PhilippWendler) [inline benchexec/tools/brick.py:34]: This is quite generic. Does the tool actually support tasks with more than one input file or no input file? If not, it would be much better for user experience if the tool-info module produced a proper error in such cases by using `task.input_files` or `task.single_input_file` as appropriate.

File: benchexec/tools/cseq.py
Comment (PhilippWendler) [inline benchexec/tools/cseq.py:46]: Please use `run.output.text` instead.

However, note that this produces different behavior than the original code: Because previously `output` contained lines with line separators, the resulting joined text would actually contain doubled line-separators (i.e., a blank line between all original lines). It is unlikely that this was intended and so the new code would actually be better, but it should be kept in mind.

Also, can the usage of `run.output.text` maybe be avoided completely? The class `RunOutput` is written specifically to provide utilities that are nice to use. For example, `if "FALSE" in output:` in this method is equivalent to `if run.output.any_line_contains("FALSE"):`.

File: benchexec/tools/divine.py
Comment (PhilippWendler) [inline benchexec/tools/divine.py:108]: Note that this changes behavior compared to the old code. Previously , `returncode` was 0 if the tool was killed by a signal, e.g., due to a timeout, now it is `None`. Is this change intended?

**Comment by skanav** [comment]
> Should I update the Ultimate tools? 

This would be great.

> How soon is it needed?

It would be needed for the preruns. So it should be done soon.

File: benchexec/tools/cbmc.py
Comment (PhilippWendler) [inline benchexec/tools/cbmc.py:112]: The fact that previously `returnsignal` and `returncode` were 0 by default is actually quite bad API and makes the code complex to understand (because for example `returncode==0` means that the tool exited successfully, but only if `returnsignal` is not `None` at the same time). In general, handling the new values (with default `None`) is easier and less errorprone. Thus I would suggest to refactor the code to directly use the fields of `run.exit_code` as they are, preferably even by passing the `ProcessExitCode` instance around instead of individual values.

File: benchexec/tools/esbmc.py
Comment (PhilippWendler) [inline benchexec/tools/esbmc.py:42]: I think this is errorprone, because the `--arch` parameter is added only if there is no item with the value `32` (or `64` respectively) in the command line. Such a value could appear as content of a different parameter, though. Furthermore, if a user adds `--arch` in their benchmark definition with a different value than what this code would produce, the parameter gets added twice.

File: benchexec/tools/symbiotic.py
Comment (PhilippWendler) [inline benchexec/tools/symbiotic.py:36]: Such a comment should not be part of the code. If you have questions, please mark the PR as draft and add the questions to the PR discussion instead.

File: benchexec/tools/symbiotic.py
Comment (PhilippWendler) [inline benchexec/tools/symbiotic.py:37]: `tool_locator` is probably better to not be stored, there is nothing in the API documentation that guarantees that this would work as expected. Furthermore, searching for the tool on disk for every call to `determine_result` is inefficient and not robust (disk content could change).

But the documentation of `BaseTool2` provides a suggestion what to do in this case already.

File: benchexec/tools/symbiotic.py
Comment (PhilippWendler) [inline benchexec/tools/symbiotic.py:111]: Note that the API documentation guarantees that this is wrong. If you want to check for empty output, this needs to be done differently.

File: benchexec/tools/two_ls.py
Comment (PhilippWendler) [inline benchexec/tools/two_ls.py:53]: Should be written as `if run.output:`

**Review by PhilippWendler — CHANGES_REQUESTED** [review]
Thanks!

I have a few comments that I added. Many of these actually apply to more than one tool-info module but I always added them in only one location, so please check for each comment in which tool-info modules it also applies.

File: benchexec/util.py
Comment (skanav) [reply benchexec/util.py:767]: That's what I was wondering too. I thought adding a separate `svcomp_utils` module in `tools` a bit of overkill.

I can add a `svcomp_util` in the `tools` package

Comment (PhilippWendler) [reply benchexec/util.py:767]: Maybe it makes more sense to add an abstract class for SV-COMP tools?

File: benchexec/tools/brick.py
Comment (skanav) [reply benchexec/tools/brick.py:34]: I have no idea if this tool or other tools with similar changes support single or multiple files . I am not a developer of this tool.
I did the changes that were upgrade the TI module without going into tool specific details.

IMO the change you suggest should be done by the tool developers.

File: benchexec/tools/symbiotic.py
Comment (skanav) [reply benchexec/tools/symbiotic.py]: fixed in fb5a18e4382326ea40bdaa4ebfb99b456f6a0ebb

File: benchexec/tools/divine.py
Comment (skanav) [reply benchexec/tools/divine.py:108]: `None != 0` also evaluates to true

File: benchexec/util.py
Comment (skanav) [reply benchexec/util.py:767]: We can add an abstract class named `CompetitionTool` because this change is relevant also for `TestComp`.

This change is is relevant for tools dealing with the language `C`, not `Java`. 
Shall we call it `CompetitionToolsForC`?

Where do you want to place this class? Separate file or `template.py`?

File: benchexec/tools/divine.py
Comment (PhilippWendler) [reply benchexec/tools/divine.py:108]: Exactly, this is why there is a difference to the previous behavior. The branch will now be entered if `value` is `None`, previously it was not entered if `value` was 0.

File: benchexec/tools/symbiotic.py
Comment (skanav) [reply benchexec/tools/symbiotic.py]: fixed in b28f883d752cc0656615bceae0bf4a5d0eeedc08

File: benchexec/tools/two_ls.py
Comment (skanav) [reply benchexec/tools/two_ls.py]: fixed in 2b4a6a1fbf69d98062f433ab690e55db316abb02

File: benchexec/util.py
Comment (PhilippWendler) [reply benchexec/util.py:767]: Hm, `CompetitionTool` or something similar would sound like this affects all kinds of competitions. In fact, its content is not even related to SV-COMP, but to SV-Benchmarks (where the `data_model` key is [defined](https://github.com/sosy-lab/sv-benchmarks#task-definitions)). So a name pointing in that direction might make sense, or something related to `C`.

We certainly should not place this class in `template.py` to make it clear it is not required by BenchExec for and tool-info modules and that it is not part of the public API. Probably we should even clearly state that out-of-tree modules should not use it and that it may change at any point in time, otherwise we would also have to version it like we do with `BaseTool2`.

**Comment by PhilippWendler** [comment]
Just remembered https://github.com/sosy-lab/sv-benchmarks/issues/1125, I think it would make sense to wait for a resolution of that issue before merging this PR.

File: benchexec/tools/brick.py
Comment (PhilippWendler) [reply benchexec/tools/brick.py:34]: Note that there are cases where the tool-info module already makes it clear that only some of these cases are supported (with an `assert` or maybe in comments). In this case the code should be cleaned up.

File: benchexec/tools/cseq.py
Comment (skanav) [reply benchexec/tools/cseq.py]: fixed in f86559fb27bf6844cf5515dba93b021247b2aeeb  for cseq and predatorHP.

File: benchexec/util.py
Comment (skanav) [reply benchexec/util.py:767]: IMO it is much simpler to put it in a util file. I will do that for now.

File: benchexec/tools/divine.py
Comment (skanav) [reply benchexec/tools/divine.py]: made it similar to others in f2d7b96e04126b075a70bf66ecc07b64b0be151e

File: benchexec/tools/brick.py
Comment (skanav) [reply benchexec/tools/brick.py]: fixed in 0af3679ba62ae7f83a8d5d7277ea12c9a35df353

File: benchexec/tools/esbmc.py
Comment (skanav) [reply benchexec/tools/esbmc.py]: fixed in fc874d47b6d1a031f0cc7403a0ebf0706670ceea

Now using the param name for searching in optipons

File: benchexec/tools/cbmc.py
Comment (skanav) [reply benchexec/tools/cbmc.py]: fixed for value in 5823bfb9ed00b07a19413ae454100020ee3a5bfa

File: benchexec/util.py
Comment (skanav) [reply benchexec/util.py:767]: @PhilippWendler do you agree with this?

Comment (PhilippWendler) [reply benchexec/util.py:767]: I would say it should have a less generic name. `toolinfo_util` sounds like an official file with officially supported API addons.

Comment (skanav) [reply benchexec/util.py:767]: do you have a suggestion?

I would like to reach an agreement before making the changes.

File: benchexec/tools/brick.py
Comment (skanav) [reply benchexec/tools/brick.py:34]: fixed in 6b526edc4561f9c86b55b7b52dbca0621a9c287b

File: benchexec/util.py
Comment (PhilippWendler) [reply benchexec/util.py:767]: `sv_benchmarks_util`?

Comment (skanav) [reply benchexec/util.py]: fixed in 14cd4fc0477b9443c6b58a4824273510afdf076e

File: benchexec/tools/brick.py
Comment (skanav) [reply benchexec/tools/brick.py:29]: I think we should wait for https://github.com/sosy-lab/sv-benchmarks/issues/1125 to be resolved before handling this.

File: benchexec/tools/map2check.py
Comment (PhilippWendler) [inline benchexec/tools/map2check.py:31]: I would prefer `_version`, AFAIK in Python it is uncommon to use `__field`.

File: benchexec/tools/sv_benchmarks_util.py
Comment (PhilippWendler) [inline benchexec/tools/sv_benchmarks_util.py:9]: Please add a note about lack of guarantees of stability and no use for out-of-tree modules (as we discussed).

File: benchexec/tools/sv_benchmarks_util.py
Comment (PhilippWendler) [inline benchexec/tools/sv_benchmarks_util.py:16]: Please add documentation. Shouldn't `data_models` be renamed because it does not primarily contain the data models, but command-line parameters (or values) for data models?

File: benchexec/tools/symbiotic4.py
Comment (PhilippWendler) [inline benchexec/tools/symbiotic4.py:54]: Can be simplified:
```suggestion
        if task.property_file:
```
Would be nice to use this consistently.

File: benchexec/tools/symbiotic4.py
Comment (PhilippWendler) [inline benchexec/tools/symbiotic4.py:69]: This is never satisfied (same holds true for condition in previous code, actually).

File: benchexec/tools/pinaka.py
Comment (PhilippWendler) [inline benchexec/tools/pinaka.py:38]: Please remove all accesses to `_lines` from all of the tool-info modules.

File: benchexec/tools/esbmc.py
Comment (PhilippWendler) [inline benchexec/tools/esbmc.py:46]: Here is another case for using `run.output.text` (without additional blank lines) or `any_line_contains` (which actually seems easy to use here).

Comment (skanav) [reply benchexec/tools/esbmc.py:46]: The tool developers use a function that they have defined themselves.
Although I understand that this could be made better but in general, I refrain from code cleanup enthusiasm for the piece of code that does not belong to me.

That was the reason I have kept it so, and also some other cases like this e.g., above comment in symbiotic.

File: benchexec/tools/symbiotic4.py
Comment (skanav) [reply benchexec/tools/symbiotic4.py]: fixed in 78f718139b1300c094474898e74de1173ee09417

File: benchexec/tools/esbmc.py
Comment (skanav) [reply benchexec/tools/esbmc.py]: fixed in 4cf9ec8e54d8a90a314ac265d5cbbbdb589e8851

File: benchexec/tools/pinaka.py
Comment (skanav) [reply benchexec/tools/pinaka.py]: fixed in af5bd81bdaf51faa9778931901fcbb1b5f251175

File: benchexec/tools/symbiotic4.py
Comment (skanav) [reply benchexec/tools/symbiotic4.py]: fixed in 66eb7e734da93d400ad5e4745dbe7b45409bf451

File: benchexec/tools/map2check.py
Comment (skanav) [reply benchexec/tools/map2check.py:31]: fixed in f2608c929d1e7ee7f86f1fb5f2674486fe00a4cb

File: benchexec/tools/sv_benchmarks_util.py
Comment (skanav) [reply benchexec/tools/sv_benchmarks_util.py:9]: fixed in 6f11a4cb9c66b01effc2b0a591ce95e368f99631

File: benchexec/tools/sv_benchmarks_util.py
Comment (skanav) [reply benchexec/tools/sv_benchmarks_util.py]: fixed in 6f11a4cb9c66b01effc2b0a591ce95e368f99631

File: benchexec/tools/cbmc.py
Comment (PhilippWendler) [inline benchexec/tools/cbmc.py:117]: I think this is buggy, right? `_prase_XML` is expected to be passed an `ProcessExitCode` instance, but gets only the return code here.

File: benchexec/tools/cbmc.py
Comment (PhilippWendler) [inline benchexec/tools/cbmc.py:140]: Lines with line separators won't ever be in `run.output`, so `\n` needs to be removed from these and other strings.

File: benchexec/tools/divine.py
Comment (PhilippWendler) [reply benchexec/tools/divine.py:107]: I don't think the current code is the intended way for this tool-info module (and behaves differently than before). For OOMs, it would return `ERROR - Pre-run` instead of `UNKNOWN` as before.

File: benchexec/tools/divine.py
Comment (PhilippWendler) [inline benchexec/tools/divine.py:110]: This is actually guaranteed to never be the case, we could remove the branch while we are at it.

File: benchexec/tools/divine4.py
Comment (PhilippWendler) [inline benchexec/tools/divine4.py:97]: This will not work, the field should be immutable.

File: benchexec/tools/klee.py
Comment (PhilippWendler) [inline benchexec/tools/klee.py:60]: This changes the semantics, because the old code had an unnoticed bug. `HARDTIMELIMIT` was never present in `rlimits`, so `--max-cputime-hard` was never passed. Now it would get added and always have the same value as `--max-cputime`.

@MartinNowack What is the intention here? BenchExec only has 2 time limits (the actual one and the hard time limit), so 2 parameters would seem to be enough? In which should the main time limit be passed?

File: benchexec/tools/legion.py
Comment (PhilippWendler) [inline benchexec/tools/legion.py:46]: I would suggest to simply return the value here instead of calling `super`. Having the full picture of what is used as command line in one method seems worth more than saving one line of code.

File: benchexec/tools/libkluzzer.py
Comment (PhilippWendler) [inline benchexec/tools/libkluzzer.py:40]: Same as above.

File: benchexec/tools/pinaka.py
Comment (PhilippWendler) [inline benchexec/tools/pinaka.py:41]: The exact semantics of the previous code would be
```suggestion
            if "VERIFICATION FAILED (ReachSafety)" in run.output:
```
(similar for the following conditions)

File: benchexec/tools/tracerx.py
Comment (PhilippWendler) [inline benchexec/tools/tracerx.py:63]: This should be handled in the same was as for Klee.

**Review by PhilippWendler — CHANGES_REQUESTED** [review]

File: benchexec/tools/libkluzzer.py
Comment (skanav) [reply benchexec/tools/libkluzzer.py:40]: fixed in 4b7464cc40a620880e02a86218a0eb1d0b26ac8c

File: benchexec/tools/legion.py
Comment (skanav) [reply benchexec/tools/legion.py:46]: fixed in 4b7464cc40a620880e02a86218a0eb1d0b26ac8c

File: benchexec/tools/divine4.py
Comment (skanav) [reply benchexec/tools/divine4.py:97]: fixed in 53d68f15f868f1dca4a307f032076a0ddc5a43c3

File: benchexec/tools/divine.py
Comment (skanav) [reply benchexec/tools/divine.py:110]: fixed in 53d68f15f868f1dca4a307f032076a0ddc5a43c3

File: benchexec/tools/cbmc.py
Comment (skanav) [reply benchexec/tools/cbmc.py:140]: fixed in a6dbfcbc3fefc99d6a7a99e9a4040b7dee88e614

File: benchexec/tools/cbmc.py
Comment (skanav) [reply benchexec/tools/cbmc.py:117]: fixed in a6dbfcbc3fefc99d6a7a99e9a4040b7dee88e614

File: benchexec/tools/pinaka.py
Comment (skanav) [reply benchexec/tools/pinaka.py]: fixed in dfb784c34463564e65652465cd7bdde773d1d174

File: benchexec/tools/divine4.py
Comment (PhilippWendler) [inline benchexec/tools/divine4.py:105]: It seems this needs the same handling as for `divine`.

File: benchexec/tools/brick.py
Comment (MartinSpiessl) [reply benchexec/tools/brick.py:32]: I find `options += [data_model_param]` way more readable than `options.append(data_model_param)`

**Comment by MartinSpiessl** [comment]
> Just remembered sosy-lab/sv-benchmarks#1125, I think it would make sense to wait for a resolution of that issue before merging this PR.

Didn't seem like there is active work on sosy-lab/sv-benchmarks#1125, but I actually depend on this PR (doesn't the whole competition infrastructure need this?), so I would welcome if this could get merged quickly.

File: benchexec/tools/divine4.py
Comment (skanav) [reply benchexec/tools/divine4.py]: fixed in 59886567a1e060133a94b09ebc336b10a5336361

File: benchexec/tools/klee.py
Comment (skanav) [reply benchexec/tools/klee.py:60]: I had a look at the `klee` wrapper submitted to sv-comp. It seems we can live with this, or even remove. Wouldn't matter much. 
This is how it is used:

```python
    time = 0
    if args.max_cputime_soft: # Use soft timeout if available
        time = args.max_cputime_soft
    elif args.max_cputime_hard:
        time = args.max_cputime_hard
    elif args.max_walltime:
        time = args.max_walltime
    elif args.max_time:
        time = args.max_time
```

Comment (PhilippWendler) [reply benchexec/tools/klee.py:60]: Good idea for looking at the script!
So they only have one time-limit internally. This should then be set to BenchExec's main time limit (previously named "soft"), not the "hard" time limit, and actually this is also what happened with the previous code.

For the new code, we could simplify this completely to
```
        if rlimits.cputime:
            options += ["--max-time=" + str(rlimits.cputime_hard)]
```
`rlimits.cputime` will always be set if there is some time limit in BenchExec, and it will be set to the value that is relevant for the tool. Passing the other values is irrelevant if the script does not use them anyway I would say.

(Of course, `--max-memory` should be kept.)

@MartinNowack Do you agree?

Comment (skanav) [reply benchexec/tools/klee.py:60]: I would say
```
        if rlimits.cputime:
            options += ["--max-cputime-soft=" + str(rlimits.cputime)]
```
because this is also what has the higher priority in the wrapper code for klee.

Comment (skanav) [reply benchexec/tools/klee.py:60]: @PhilippWendler wouldn't BenchExec anyways kill the process at hard time limit?

Comment (PhilippWendler) [reply benchexec/tools/klee.py:60]: Would also be fine for me, @MartinNowack would be best to decide this. If he does not respond, we can still do this.

Comment (PhilippWendler) [reply benchexec/tools/klee.py:60]: > wouldn't BenchExec anyways kill the process at hard time limit?

Yes, this is what happens.

Comment (skanav) [reply benchexec/tools/klee.py]: fixed in aaeae7166b8469eebf828b91ee9f6fe768e1afc7

if @MartinNowack prefers some other flag, it can be changed later.

File: benchexec/tools/brick.py
Comment (skanav) [reply benchexec/tools/brick.py:29]: @PhilippWendler now I wonder if we should wait for https://github.com/sosy-lab/sv-benchmarks/issues/1125 to be resolved, or just go ahead with this and incorporate the changes later.

Comment (PhilippWendler) [reply benchexec/tools/brick.py:29]: I would say that refactoring the constants out into the utility module should be done, because it might make adjusting the tool-info modules to the result of sosy-lab/sv-benchmarks#1125 easier.

But we could avoid blocking this PR on sosy-lab/sv-benchmarks#1125 if there is no progress there.

**Comment by MartinNowack** [comment]
From my side, changes for KLEE look good.

**Comment by PhilippWendler** [comment]
@MartinNowack  Thanks for confirmation!

**Comment by skanav** [comment]
@PhilippWendler do we need something more here?

**Review by PhilippWendler — APPROVED** [review]

**Comment by PhilippWendler** [comment]
I had just hoped that we get some response for sosy-lab/sv-benchmarks#1125, but as discussed let's merge it now.

**Comment by PhilippWendler** [comment]
@skanav I just [noticed](https://gitlab.com/sosy-lab/sv-comp/archives-2021/-/jobs/826667408) that this broke some other tool-info modules that inherit from modules changes in this PR. At least JBMC is affected. Would it be possible for you to have a look at which tool-info modules inherit from the ones changed here and adjust them (this should typically be small changes only)?

**Pull request merged**
